server/util/schema.ts

```diff
@@ -1,4 +1,52 @@
+import { ListQueryParams, Order, orders } from '../../shared';
+
 export function AssertKeySchema(keys: string[] | null | undefined, schema: readonly string[]) {
-	const error = new Error(`provided keys '${keys?.join(' ')}' do not match the required '${schema.join(' ')}' schema!`);
-	if (!keys || keys.join() !== schema.join()) throw error;
+  const error = new Error(`provided keys '${keys?.join(' ')}' do not match the required '${schema.join(' ')}' schema!`);
+  if (!keys || keys.join() !== schema.join()) throw error;
+}
+
+export class InvalidType extends Error {
+  constructor(field: string, value: unknown, typename: string) {
+    super(`Invalid \`${field}\`; \`${value}\` is not of type \`${typename}\``);
+  }
+}
+
+const DEFAULT_ORDER = 'asc';
+
+/**
+ *
+ * @param params params to extract
+ * @param filterable values that can be filtered by
+ * @param sortable values that can be sorted by
+ * @returns typed and validated query params
+ */
+export function getListParams<T extends Record<string, unknown>>(
+  params: Record<string, unknown>,
+  filterable: (keyof T)[],
+  sortable: (keyof T)[],
+) {
+  const query: ListQueryParams = {};
+  const { filter, for: _for, sort } = params;
+
+  if (filter) {
+    if (typeof filter !== 'string') throw new InvalidType('filter', filter, 'string');
+    if (!filterable.includes(filter)) throw new Error(`Cannot filter by \`${filter}\`'`);
```

Comment:
there is an extra quote mark in you error here (sorry, a bit picky)
![image](https://user-images.githubusercontent.com/83617997/227895937-d7216d2d-9bc8-4927-af16-cf5af8642f94.png)